compiler/optimizations/remoteValueForwarding.cpp

```diff
@@ -84,6 +88,83 @@ void remoteValueForwarding() {
   }
 }
 
+static void fixLHSArray(CallExpr* move, std::vector<Symbol*>& todo) {
+  Symbol* LHS = toSymExpr(move->get(1))->symbol();
+  if (LHS->isRef()) {
+    LHS->type = LHS->getValType();
+    LHS->qual = QUAL_VAL;
+    todo.push_back(LHS);
+  }
+}
+
+static void replaceRecordWrappedRefs() {
+  std::vector<Symbol*> todo;
+  forv_Vec(AggregateType, aggType, gAggregateTypes) {
+    if (!aggType->symbol->hasFlag(FLAG_REF)) {
+      for_fields(field, aggType) {
+        if (field->isRef() && isRecordWrappedType(field->getValType())) {
+          field->type = field->getValType();
+          field->qual = QUAL_VAL;
+          todo.push_back(field);
+        }
+      }
+    }
+  }
+
+  // I'd like to be able to just iterate over the uses of tuple fields, but
+  // we don't have a good way of doing that today. The case to worry about
+  // is when we're indexing into a tuple with an integer (param or otherwise).
+  forv_Vec(CallExpr, call, gCallExprs) {
+    if (call->isPrimitive(PRIM_GET_SVEC_MEMBER_VALUE)) {
+      CallExpr* move = toCallExpr(call->parentExpr);
+      INT_ASSERT(isMoveOrAssign(move));
+
+      if (!call->isRef()) {
+        INT_ASSERT(isRecordWrappedType(call->typeInfo()->getValType()));
+        fixLHSArray(move, todo);
+      }
+    }
+  }
+
+  // Try and find references that were initialized with the field reference,
+  // and fix them to be QUAL_VAL
+  while (todo.size() > 0) {
+    Symbol* sym = todo.back();
+    todo.pop_back();
+    INT_ASSERT(!sym->isRef() && isRecordWrappedType(sym->type));
+
+    for_SymbolSymExprs(se, sym) {
+      CallExpr* call = toCallExpr(se->parentExpr);
+      INT_ASSERT(call);
+
+      if (call->isPrimitive(PRIM_MOVE)) {
```

Comment:
How do you know that you only need to handle PRIM_MOVE/PRIM_GET_MEMBER_VALUE/PRIM_RETURN here?
If it's just because those are the only current AST patterns that need repair, maybe add a comment to that effect? Or, even better, think about some other cases?

What if the field was passed by reference to a function? Don't you need to add an Addr-Of now? Should this pass also be changing record-wrapped arguments to be passed with 'in' intent, now that copying is handled? (ie instead of passing ref(_array), pass _array since the _array record itself is immutable).